### ISSUE TITLE
Feature request: parallel tests

### ISSUE THREAD

**Issue by Tux** [briandfoy/module-release#38]
Currently testing Text::CSV_XS with 145+ versions of perl takes more than an hour.

Most of the consumed time is running the tests (52000+ tests taking 15-20 seconds)

I can see two approaches for a speedup.

1: if the t/ folder is parallel-safe, use -jN
```
$ prove -wbQ t
All tests successful.
Files=29, Tests=52231, 17 wallclock secs ( 3.12 usr  0.20 sys + 16.47 cusr  0.47 csys = 20.26 CPU)
Result: PASS
```
```
$ prove -wbQ -j8 t
All tests successful.
Files=29, Tests=52231, 12 wallclock secs ( 3.67 usr  0.22 sys + 18.01 cusr  0.67 csys = 22.57 CPU)
Result: PASS
```
```
$ make test
:
All tests successful.
Files=29, Tests=52231, 17 wallclock secs ( 3.20 usr  0.23 sys + 16.16 cusr  0.42 csys = 20.01 CPU)
Result: PASS
```
```
$ env HARNESS_OPTIONS=j8 make test
:
All tests successful.
Files=29, Tests=52231, 12 wallclock secs ( 3.77 usr  0.25 sys + 18.70 cusr  0.50 csys = 23.22 CPU)
Result: PASS
```
Implementing this is easy:
```
--- release     2021-03-11 08:56:52.000000000 +0100
+++ release-2   2021-03-24 11:19:51.325116916 +0100
@@ -35,6 +35,10 @@ release - give your Perl distros to the
        # set $ENV{AUTOMATED_TESTING} to a true value
        release -a

+       # Enable parallel testing by setting $ENV{HARNESS_OPTIONS}
+       release -j
+       release -j8
+
 =head1 DESCRIPTION

 This is the prototype program for using C<Module::Release>. You should
@@ -321,8 +325,18 @@ my @extra_opts = split /\s+/, $ENV{RELEA
 push @ARGV, @extra_opts;

 my %opts;
-getopts('aChdptvVkT', \%opts) or $opts{h} = 1;
+getopts('aChdptvVkTj:', \%opts) or $opts{h} = 1;

+if( exists $opts{j} ) {
+       unless( defined $opts{j} ) {
+               delete $opts{h}; # getopts does not support optional arguments like Getopt::Long does in "j|parallel:9"
+               $opts{j} = eval {
+                       require System::Info;
+                       System::Info->new->get_core_count;
+                       } || 9;
+       }
+       $ENV{HARNESS_OPTIONS} = "j$opts{j}";
+}

 # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
 my( $script_version ) = '2.127';
```
setting this option reduces my `release -t` on my laptop today from 65 minutes to 52 minutes. Not a huge win, but still.

2: run tests in parallel in different cloned folders
I envision more gain if I could run in a different folder per perl version 8 (or whatever) folders with a clone in parallel. This would however mean a lot more work.

If I pass a base folder like `~/test/tmp` where the current folder is rsync'd too before running the tests and than clean-up when done, I can run 8 perl versions in parallel and use more of the available resources.

**Comment by briandfoy** [comment]
Do you really need release to set an environment variable for you? I guess I've already done that with `-a` though.

Can you make this a pull request?

**Comment by briandfoy** [comment]
Also, I think these options should also be available in the config, so I'll have to think about that too. It's not just this one, but others as well. That's not necessary to implement this as `-j` though.

**Comment by Tux** [comment]
It was more or less a braindump that I wrote while `release -t` was purring along. Analysing the code made me come up with the ideas.
I always forget the syntax for `HARNESS_OPTIONS`. Most often I try `env TEST_JOBS=8 ...` first, only to notice that this is perl CORE only, and I have ti dig up the Test::Harness manual again. I however never forget that `prove` and `make` both accept `-j`, so it is just a matter of convenience.
I'll make a PR for the `-j` diff.
I didn't try to implement the second plan (would have been `-J[#]`) as time is not on my side right now, but I seriously consider making a more serious attempt.
I'll also have a look at the config: that makes good sense.

**Comment by briandfoy** [comment]
I've done the config stuff in #39, so just do the same for `-j`. If the config setting is 0, HARNESS_OPTIONS is not changed.

**Comment by briandfoy** [comment]
Also, think about how to handle a HARNESS_OPTIONS that is already set. We don't want other settings to disappear. Something like:

     $ENV{HARNESS_OPTIONS} .= " j$opts{j}";

Or maybe something like this, perhaps replacing any `j` option that is already there.

    $ENV{HARNESS_OPTIONS} = join ' ', split( /\s+/, $ENV{HARNESS_OPTIONS} ), "j$opts{j}"

**Comment by Tux** [comment]
Noted. Won't be today: working on a Text::CSV_XS release, and Z/OS is slow (but the tests PASSed!)

**Comment by Tux** [comment]
Done. PR sent

**Comment by briandfoy** [comment]
The PR is #40.

Some things I want to look at in master now that it is merged:

- [ ] some inconsistent whitespace (tabs to indent, spaces to align)
- [ ] exit code for usage()
- [ ] Need to update usage message for all switches
- [ ] Check other docs to ensure all the switches are listed
- [ ] I think I have some doubled up debug messages lying around

**Comment by briandfoy** [comment]
Cleaned up in 0cd88d717e3e43011c1e6623877b620a1692cbe3
